File: day09/IntCode.js
class IntCode {
    constructor(name, program, input = null, debug = false) {
        this.name = name;
        this.input = input;
        this.program = [...program];
        // add a bunch of zeroes to the program array
        for (let i = 0; i < 1000; i++) {
            this.program.push(0);
        }
        this.i = 0;
        this.running = true;
        this.waitingForInput = false;
        this.output = null;
        this.debug = debug;
        this.relativeBase = 0;
        this.opcode = null;
        this.a = null;
        this.b = null;
        this.c = null;
        this.run();
    }

    run() {
        while (this.running && !this.waitingForInput) {
            if (this.program[this.i] === 99) {
                if (this.debug) {
                    console.log('exiting');
                }
                this.exited = true;
                return 99;
            }

            this.getInstruction();
            if (this.debug) {
                console.log(this.name, 'opcode: ', this.opcode);
            }

            switch (this.opcode) {
                case 1: // add
                    this.add();
                    break;
                case 2: // mul
                    this.mul();
                    break;
                case 3: // input
                    if (this.input !== null) {
                        this.processInput(this.input);
                        this.input = null;
                    } else {
                        this.waitingForInput = true;
                        return this.opcode;
                    }
                    break;
                case 4: // output
                    this.printOutput();
                    break;
                case 5: // jump if true
                    this.jumpIfTrue();
                    break;
                case 6: // jump if false
                    this.jumpIfFalse();
                    break;
                case 7: // less than
                    this.lessThan();
                    break;
                case 8: // equal
                    this.equalTo();
                    break;
                case 9: // adjust relative base
                    this.adjustRelativeBase();
                    break;
                default:
                    throw new Error(`unknown opcode ${this.opcode}`);
            }
        }

        throw new Error(`error running program`);
    }

    add() {
        if (this.debug) {
            console.log(`adding ${this.a} + ${this.b} (${this.a + this.b}) into index ${this.c}`);
        }
        this.program[this.c] = this.a + this.b;
        this.i += 4;
    }

    mul() {
        if (this.debug) {
            console.log(`mul ${this.a} * ${this.b} (${this.a * this.b}) into index ${this.c}`);
        }
        this.program[this.c] = this.a * this.b;
        this.i += 4;
    }

    processInput(input) {
        if (this.debug) {
            console.log(`${this.name}: placing input of ${input} into index ${this.a}`);
        }
        this.program[this.a] = input;
        this.i += 2;
        this.waitingForInput = false;
    }

    printOutput() {
        if (this.debug) {
            console.log(`printing from index ${this.a}`);
        }
        console.log('PROGRAM OUTPUT: ', this.a);
        this.output = this.a;
        this.i += 2;
    }

    adjustRelativeBase() {
        if (this.debug) {
            console.log(`Adding ${this.a} to relative base ${this.relativeBase}: ${this.relativeBase + this.a}`);
        }
        this.relativeBase += this.a;
        this.i += 2;
    }

    jumpIfTrue() {
        if (this.a !== 0) {
            if (this.debug) {
                console.log(`${this.a} is non zero, jumping to ${this.b}`);
            }
            this.i = this.b;
        } else {
            if (this.debug) {
                console.log(`${this.a} is zero - skipping`);
            }
            this.i += 3;
        }
    }

    jumpIfFalse() {
        if (this.a === 0) {
            if (this.debug) {
                console.log(`${this.a} is zero, jumping to ${this.b}`);
            }
            this.i = this.b;
        } else {
            if (this.debug) {
                console.log(`${this.a} is non zero - skipping`);
            }
            this.i += 3;
        }
    }

    lessThan() {
        if (this.a < this.b) {
            if (this.debug) {
                console.log(`${this.a} is less than ${this.b}, storing 1 into index ${this.c}`);
            }
            this.program[this.c] = 1;
        } else {
            if (this.debug) {
                console.log(`${this.a} is not less than ${this.b}, storing 0 into index ${this.c}`);
            }
            this.program[this.c] = 0;
        }
        this.i += 4;
    }

    equalTo() {
        if (this.a === this.b) {
            if (this.debug) {
                console.log(`${this.a} equals ${this.b}, storing 1 into index ${this.c}`);
            }
            this.program[this.c] = 1;
        } else {
            if (this.debug) {
                console.log(`${this.a} is not equal to ${this.b}, storing 0 into index ${this.c}`);
            }
            this.program[this.c] = 0;
        }
        this.i += 4;
    }

    getInstruction() {
        const instruction = this.program[this.i];
        // const result = [instruction];
        this.opcode = instruction;

        const paramModes = [0, 0, 0];
        if (instruction > 9) {
            const opcodeDigits = instruction.toString().split('').reverse();
            this.opcode = Number(opcodeDigits[1] + opcodeDigits[0]);
            paramModes[0] = opcodeDigits[2] ? Number(opcodeDigits[2]) : 0;
            paramModes[1] = opcodeDigits[3] ? Number(opcodeDigits[3]) : 0;
            paramModes[2] = opcodeDigits[4] ? Number(opcodeDigits[4]) : 0;
        }

        if (paramModes[0] === 0) {
            // position mode - param is pointer
            if (this.opcode !== 3) {
                this.a = this.program[this.program[this.i + 1]];
            } else {
                this.a = this.program[this.i + 1];
            }
        } else if (paramModes[0] === 1) {
            // immediate mode - param is value
            this.a = this.program[this.i + 1];
        } else if (this.opcode !== 3) {
            // relative mode - param is pointer + relative base
            this.a = this.program[this.program[this.i + 1] + this.relativeBase];
        } else {
            this.a = this.program[this.i + 1] + this.relativeBase;
        }

        if (paramModes[1] === 0) {
            // position mode
            this.b = this.program[this.program[this.i + 2]];
        } else if (paramModes[1] === 1) {
            // immediate mode
            this.b = this.program[this.i + 2];
        } else {
            // relative mode
            this.b = this.program[this.program[this.i + 2] + this.relativeBase];
        }

        if (paramModes[2] === 0) {
            this.c = this.program[this.i + 3];
        } else if (paramModes[2] === 2) {
            this.c = this.program[this.i + 3] + this.relativeBase;
        } else if (paramModes[2] === 1) {
            throw new Error('implement immediate mode for c????');
        }
    }
}

export default IntCode;

File: day13/intcode.py
ADD, MUL, IN, OUT, JNZ, JZ, LT, EQ, REL = range(1, 10)
HALT = 99

OPCODE_NUM_PARAMS = {
    ADD: 3,
    MUL: 3,
    IN: 1,
    OUT: 1,
    JNZ: 2,
    JZ: 2,
    LT: 3,
    EQ: 3,
    REL: 1,
    HALT: 0,
}

MODE_POSITION, MODE_IMMEDIATE, MODE_RELATIVE = range(3)
TYPE_SRC, TYPE_DEST = range(2)

OPCODE_PARAM_TYPES = {
    ADD: (TYPE_SRC, TYPE_SRC, TYPE_DEST),
    MUL: (TYPE_SRC, TYPE_SRC, TYPE_DEST),
    IN: (TYPE_DEST,),
    OUT: (TYPE_SRC,),
    JNZ: (TYPE_SRC, TYPE_SRC),
    JZ: (TYPE_SRC, TYPE_SRC),
    LT: (TYPE_SRC, TYPE_SRC, TYPE_DEST),
    EQ: (TYPE_SRC, TYPE_SRC, TYPE_DEST),
    REL: (TYPE_SRC,),
    HALT: 0,
}


class Intcode:
    def __init__(self, code) -> None:
        self.program = code[:] + [0] * 10000
        self.pointer = 0
        self.running = True
        self.output = []
        self.relative_base = 0

    def run(self, input=None):
        self.output = []
        while self.running is True:
            opcode = self.program[self.pointer]

            if opcode == 99:  # exit
                self.running = False
                return self.output

            modes = (
                (opcode // 100) % 10,
                (opcode // 1000) % 10,
                (opcode // 10000) % 10,
            )
            opcode %= 10
            num_params = OPCODE_NUM_PARAMS[opcode]
            types = OPCODE_PARAM_TYPES[opcode]
            params = self.program[self.pointer + 1 : self.pointer + 1 + num_params]

            for i in range(len(params)):
                if modes[i] == MODE_POSITION:
                    if types[i] == TYPE_SRC:
                        params[i] = self.program[params[i]]
                elif modes[i] == MODE_RELATIVE:
                    if types[i] == TYPE_SRC:
                        params[i] = self.program[self.relative_base + params[i]]
                    elif types[i] == TYPE_DEST:
                        params[i] += self.relative_base

            if opcode == ADD:  # add
                a, b, c = params
                self.program[c] = a + b
                self.pointer += 4
            elif opcode == MUL:  # mul
                a, b, c = params
                self.program[c] = a * b
                self.pointer += 4
            elif opcode == IN:  # input
                a = params[0]
                if input is not None:
                    self.program[a] = input
                    self.pointer += 2
                    input = None
                else:
                    return self.output
            elif opcode == OUT:  # output
                a = params[0]
                # print("OUTPUT: ", a)
                self.output.append(a)
                self.pointer += 2
            elif opcode == JNZ:  # jump if true
                a, b = params
                self.pointer = b if a != 0 else self.pointer + 3
            elif opcode == JZ:  # jump if false
                a, b = params
                self.pointer = b if a == 0 else self.pointer + 3
            elif opcode == LT:  # less than
                a, b, c = params
                self.program[c] = 1 if a < b else 0
                self.pointer += 4
            elif opcode == EQ:  # equal
                a, b, c = params
                self.program[c] = 1 if a == b else 0
                self.pointer += 4
            elif opcode == REL:  # adjust relative base
                a = params[0]
                self.relative_base += a
                self.pointer += 2
            else:
                print(f"Unknown opcode: {opcode}")
                return None
